### ISSUE TITLE
PROPOSAL - Open standup window

### ISSUE THREAD

**Issue by hubpav** [standup-raven/standup-raven#16]
We are implementing standup bot for the whole company - however, many people complain they cannot fill up their standup tasks in the evening (we do regular reports at 9 am). Raven Bot does not allow to set the starting window earlier than the closing window (so there could be like 9 pm till 9 am time frame).

It would be nice to:

1. Allow wide time frames from the previous day
2. Allow the possibility to disable the standup window whatsoever - so anything after report time (e.g. 9 am) will be aggregated for the next day ... it can be done in a way when the report window is set 0:00-0:00

What do you think?

**Comment by harshilsharma** [comment]
Thanks for the proposal @hubpav .

The way we manage reports and standups right now doesn't allow for a single standup window to span more than one day.

I suppose if we allowed submitting standup for the next day, folks could submit the standup in the evening and the reports include them the next day. Would this fit your scenario?

**Comment by hubpav** [comment]
@harshilsharma I believe you got it right. The point is to allow standup submission in the evening. I have not checked the implementation and what are the options. We don't need the standup window to span multiple days, just the flexibility to go ahead right after the prior standup finished (intentionally I am providing an extreme use-case).

Does this sound feasible to you? Thanks!

**Comment by harshilsharma** [comment]
@hubpav I understand the general feature here - ability to fill future standups. We want to explore if others would be interested in the feature as well. I'll be checking on that and if the community likes it, I don't see a problem with that!

**Comment by gcerice** [comment]
@harshilsharma 
We have a similar need.  Our team has members in US & India.  We need US contributions made today included in the same report with contributions from India tonight.   We have our daily standup phone calls at 5:45am US and would set the Raven window to end just shy of then.  However we need the Raven window to begin in the US afternoon of the previous day.   We could ideally configure the plugin to begin at 3:30pm US and end at 5:30am US, which would take into account end of US workday and most of India workday.

**Comment by harshilsharma63** [comment]
@gcerice I understand your requirements and would to work this out.

**Comment by gcerice** [comment]
@harshilsharma63 disregard my input. We found work-around by setting up the raven window using India timezone instead of US timezone. Within IST, both start and stop times occur before midnight.
4:30a IST --> 5:00p IST then becomes 5:00p CST --> 5:30a CST

**Comment by Killerbear** [comment]
we would also love to be able to set a starting time that is "after" the ending time.
If the starting time is after the ending time would just mean, that ending time is on the next day.

our usecase is as follows:
daily meeting at 10 o'clock.
and everything added to the raven after this meeting should be added to the daily for the next day.
our prefered time windows would then start: 11:00 and end: 10:00 (next day)

currently the raven complains:
"An error occurred while saving standup config.
Window open time cannot be after window close time"

**Comment by monsdar** [comment]
This is the first thing that came up after we set up the plugin within our Mattermost instance. Would be nice to be able to fill in the next standup anytime after the previous has passed.

Multiple teams asked for this individually, so it seems this would be an intuitive thing to have.

**Comment by harshilsharma63** [comment]
@Killerbear @monsdar  why we cannot treat anything filled after standup report as next standup is because sometimes when people miss filling their standup, they fill it after the report and re-generate. the report.

The simplest solution I can think of is adding a datepicker to standup submission modal. that allows you to select the date you want to submit the standup for. This approach keeps the standup scheduling logic simple and as is and allows one to submit standup not just. for. the next. day but any day in advance.

How does that sound to you @Killerbear @monsdar ?

**Comment by Killerbear** [comment]
@harshilsharma this sounds like a good solution.
Important is for everyone else is that you should not need to select a date everytime you want to submit to the raven. Only if you want to post for a different date.

This means that nothing should change for the common usage of the raven.
The default day selected in the datepicker should be the current day.
And only for our usecase you can click on the datepicker and change the day.

**Comment by harshilsharma63** [comment]
@Killerbear yes, agreed. Would you be interested in working on this feature?

**Comment by Killerbear** [comment]
@harshilsharma
i havent work with react or go before. but i can try.
dont expect too much ^^

**Comment by harshilsharma63** [comment]
@Killerbear hey sure, here to help if you face any issue. Do you have any approach on mind?

**Comment by Killerbear** [comment]
@harshilsharma my current ideas are:

1. put a datepicker on every input field
2. add a button at the buttom where Cancel/Submit is with "change submit date" -> opens a datepicker
3. add a button at the buttom "add report" -> opens datepicker -> after picking a date a new report form is shown

my favorite one is second.
If the raven is empty you can change the date and see the datepicker. already used dates are highlighted. and if you click on that date the submit of that date gets loaded into the raven.

if the raven is filled and you click on another date that already has a submit ... then i dont know ... i have to think about that case.

**Comment by harshilsharma63** [comment]
@Killerbear I like the second option as well but not sure if bottom is the right place for this. I was thinking of a next and previous day buttons on top itself for quick navigation to next day. And a button for selecting. the date like you've mentioned.

I really like the idea of highlighting the days for which a standup is submitted 👍 

> if the raven is filled and you click on another date that already has a submit ... then i dont know ... i have to think about that case.

I think we'll have to show a. confirmation dialog to indicate that unsaved data will be lost. And they can continue or cancel.

I can create some designs in Figma for date picker if that helps.

**Comment by Killerbear** [comment]
@harshilsharma the idea with the confirmation dialog is good. 👍

i can also put the buttons on top but
i wanted to made this feature as less important as possible because 99% of the users will never use this feature (i assume). So i would like to add only one single date button.
Because having 3 buttons permanently cluttering the UI but will never be used by the majority of the users is maybe not a good decision.

Imho it would be a good choice to make the date selection a subtle available option with one date button. This is obviously not the quickes way to navigate through the days but because you probably are not gonna use this feature every day, it will be okay.

the exact position for the button is still open.
i can also make some quick mockups in Photoshop.

my currently problem is how to get the standup raven running for development.
the dev guide only shows how to download and build the app. but how do i run it?
do i need a mattermost server for development?

**Comment by harshilsharma63** [comment]
@Killerbear agreed with your point of making it subtle. It'll definitely be used by more than 1% users (make 10%?) as we've received this feature request multiple times.

> do i need a mattermost server for development?

Yes. Since this a Mattermost plugin, you need a Mattermost server running on your system.
You can get one running following the official Mattermost guide available at https://mattermost.com/download/ .

**Comment by monsdar** [comment]
I just found that using Docker to develop plugins works in a good way with Mattermost. Perhaps it's a bit easier to set up. Documented my findings [here](https://yetiblog1337.wordpress.com/2020/06/21/writing-mattermost-plugins-with-vscode-and-docker/)

**Comment by harshilsharma63** [comment]
@Killerbear hi there! Did you get a chance to think about the mockups?

**Comment by Killerbear** [comment]
@harshilsharma i was trying to get the raven up and running on my machine.
But the readme isnt really helpful especially for people on Windows.
Instead i was trying to follow the link that monsdar posted 2 comments above cause i like the idea of using docker for development.
But it gets stuck on the `make` process.

Maybe you're right and i should start with the mockups first. ^^

**Comment by harshilsharma63** [comment]
> I was trying to get the raven up and running on my machine. But the readme isnt really helpful especially for people on Windows.

@Killerbear are you talking about Mattermost's readme?

Have you tried these steps for running Mattermost on Windows with Docker - https://developers.mattermost.com/contribute/server/developer-setup/ ?

Can  you post the output when you are stuck? You can also post your query at https://community.mattermost.com/core/channels/developers . I'm sure someone else has faced the same issue as well.

**Comment by monsdar** [comment]
Make sure to increase your memory limit for the docker containers. Build got stuck for me as well when i was giving only 2gb. Solved it by giving 8gb.

**Comment by harshilsharma63** [comment]
@Killerbear one of my friends suggested installing Windows Subsystem for Linux then running Mattermost directly from the source code.

**Comment by monsdar** [comment]
Or try out docker for Windows with WSL2 backend. It's on my bucket list as well.

**Comment by Killerbear** [comment]
> Make sure to increase your memory limit for the docker containers. Build got stuck for me as well when i was giving only 2gb. Solved it by giving 8gb.

i've already done this. maybe i can try even more.

> @Killerbear one of my friends suggested installing Windows Subsystem for Linux then running Mattermost directly from the source code.
> Or try out docker for Windows with WSL2 backend. It's on my bucket list as well.

yes i can try this.

i can also try cygwin
or installing ubuntu on a new partition

we'll see how it goes. ^^

**Comment by harshilsharma63** [comment]
@Killerbear Windows Subsystem for Linux is far superior to Cygwin or a VM. WSL integrates better, is actively maintained, and works just so well than the other two solutions.

**Comment by Killerbear** [comment]
@harshilsharma63 i was curious about Figma so i gave it a try.
Pretty nice tools. works very well in the browser.
here is how i would imagine it.
https://www.figma.com/file/n6MOG0xehggDoHQeiSIhbR/Standup-Raven?node-id=0%3A1

maybe the icon and text needs to be a little brighter but i think the position on the first one is pretty intuitive.
The order of actions is: 1. Fill out the form -> 2. change date if necessary -> 3. Submit
In that case the datepicker needs to be after the inputs and before the submit button.
And in western countries the way of attention is from top to bottom and from left to right.
![image](https://user-images.githubusercontent.com/5721655/86519342-2f598d00-be3a-11ea-900b-c7a9fae53996.png)

Which means 1. inputs -> 2. datepicker -> 3. Cancel/Submit Buttons
And this correlates to the order of actions.

**Comment by harshilsharma63** [comment]
@Killerbear screenshots look neat! I don't think I have permission for the Figma file. You'll need to generate a sharing URL.

**Comment by Killerbear** [comment]
@harshilsharma63 i changed the permissions now. you should be able to see something.

**Comment by harshilsharma63** [comment]
@Killerbear that looks neat 👌. We already have a react-bootstrap date picker in the plugin so all we need to do here in the UI department is add the date picker to standup modal.

**Comment by Killerbear** [comment]
just for a quick update. i'm still trying to get this up and running. WSL is pretty nice but its not solving the problems. i was hoping that using docker would solve some problems in setting up mattermost but instead it just creates new problems. but yeah. i'm still on it.

**Comment by laurNtiu** [comment]
This feature would be extremely useful for weekly stand-ups. Right now we notice that reports get assigned to days where no stand-up is schedule. E.g. we have a stand-up on Monday, but people are filling their list on Friday, so we need to generate reports for every day when someone may have filled in their reports; not ideal.

Another idea for solving thisfor weekly meetings is to allow multi-day windows for submitting reports. You would only  need to switch from specifying hours to an "All-day" switch + number of days in advance input box. E.g. for a weekly Monday meeting, I would fill in All-day = true, AdvancedDays = 5, equivalent to Thu-Mon. Might be easier to implement.
